src/thread.cpp

```diff
@@ -22,20 +24,149 @@
 #if defined(_WIN32)
 #define WIN32_LEAN_AND_MEAN 1
 #include <windows.h>
-#include <cstdlib> // mbstowcs
+#include <condition_variable>
+#include <cstdlib>  // mbstowcs
+#include <mutex>
+#include <vector>
 #elif defined(__APPLE__)
 #include <mach/thread_act.h>
 #include <pthread.h>
 #include <unistd.h>
+#include <thread>
 #else
 #include <pthread.h>
 #include <unistd.h>
+#include <thread>
 #endif
 
 namespace marl {
 
 #if defined(_WIN32)
 
+#define CHECK_WIN32(expr)                                    \
+  do {                                                       \
+    auto res = expr;                                         \
+    (void)res;                                               \
+    MARL_ASSERT(res == TRUE, #expr " failed with error: %d", \
+                (int)GetLastError());                        \
+  } while (false)
+
+namespace {
+
+struct ProcessorGroup {
+  unsigned int count;  // number of logical processors in this group.
+  KAFFINITY affinity;  // affinity mask.
+};
+
+const std::vector<ProcessorGroup>& getProcessorGroups() {
+  static std::vector<ProcessorGroup> groups = [] {
+    std::vector<ProcessorGroup> out;
+    SYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX info[32] = {};
```

Comment:
I couldn't see anything in the spec, but I assume this won't go over 32?
(Looks like there are only 5 Relationships, so I wouldn't expect it, but could you have the same relationship more than once, for example multiple physical CPUs?)

Reply:
I also couldn't see anything in the spec, and I assumed I'd only need 1 as I'm requesting 1 relationship. What I saw was 1 filled in entry, another with a relationship id that contained a magic number that was not in the enum, nor documented anywhere. I assume this is an EOS marker. 32 is a magic number, but it is larger than I think we need, I have no documentation to use anything larger or smaller, and I'm passing the size to the call that fills it. So *shrug*